chore(components): add development documentation (#4493)
Automattic/newspack-plugin: packages/components/DEVELOPMENT.md

## Contents

- [Purpose](#purpose)
- [When to Use](#when-to-use)
- [Component Selection Hierarchy](#component-selection-hierarchy)
- [Design & layout at a glance](#design--layout-at-a-glance)
- [Available Components](#available-components)
- [Usage](#usage)
- [Import Patterns](#import-patterns)
- [Common WordPress Components](#common-wordpress-components-used-alongside-newspack-components)
- [Styling](#styling)
- [Component Patterns](#component-patterns) *(for contributors)*
- [Component Development Guidelines](#component-development-guidelines)
- [Related Packages](#related-packages)
- [Testing](#testing)

## Purpose

This package provides custom React components designed specifically for Newspack backend/admin interfaces (wizards, settings pages, etc.). These components are built on top of WordPress components and provide Newspack-specific functionality, styling, and patterns.

> **In short:** For backend/admin screens use Newspack components first (Card, ActionCard, SectionHeader, etc.), fall back to WordPress components when needed, and follow the spacing scale and hierarchy patterns so UIs stay consistent. For blocks use WordPress components (and `AutocompleteTokenField` only when you need autocomplete). For reader-facing UI use Newspack UI or theme components, not this package.

Reviewer: ```suggestion
**In short:** For backend/admin screens use Newspack components first (Card, ActionCard, SectionHeader, etc.), fall back to WordPress components when needed, and follow the spacing scale and hierarchy patterns so UIs stay consistent. For block editor UI use WordPress components (and `AutocompleteTokenField` only when you need autocomplete). For reader-facing UI use Newspack UI or theme components, not this package.
```

**Design-wise:** Backend UIs should feel consistent with the WordPress admin, with clear visual hierarchy (section → card → controls) and predictable spacing. Use the same components and spacing scale across wizards so design and code stay aligned; when introducing a new pattern or layout, align with design (and designer review) before implementing.

## When to Use

### Decision Tree

**Backend/Admin UI (Wizards, Settings Pages):**
- ✅ **Use Newspack components first** - Check if a Newspack component exists for your use case
- ⚠️ **Fallback to WordPress components** - If no Newspack component exists, use `@wordpress/components`
- Examples: Dashboard, Settings, Audience Management, Setup Wizard

**Gutenberg Blocks:**

> - ✅ **Use WordPress components** - Blocks should use `@wordpress/components` for consistency with the block editor

Reviewer: ```suggestion
- ✅ **Use WordPress components** - Block editor UI should use `@wordpress/components` for consistency with the block editor
```

- ⚠️ **Exception:** `AutocompleteTokenField` - Can be used in blocks when autocomplete functionality is needed
- Examples: All blocks in `src/blocks/` use WordPress components

> **Frontend (Reader-facing UI):**

Reviewer: ```suggestion
**Frontend (Reader-facing UI, including block render output):**
```

- ❌ **Don't use Newspack components** - Use Newspack UI (`src/newspack-ui/`) or theme components instead
- Examples: My Account, Reader Activation modals, Auth screens

## Component Selection Hierarchy

Follow this step-by-step process when selecting a component:

1. **Check Newspack components first**
   - Review the [Available Components](#available-components) list below
   - Check `packages/components/src/` directory for component implementations
   - Newspack components are optimized for backend/admin workflows

2. **If not available, use WordPress components**
   - Visit [@wordpress/components Storybook](https://wordpress.github.io/gutenberg/?path=/docs/docs-introduction--page)
   - Check [@wordpress/components npm package](https://www.npmjs.com/package/@wordpress/components)

>    - WordPress components provide standard admin UI patterns

Reviewer: Since sometimes new components are introduced, or component API changes across package versions:

```suggestion
   - Cross-reference Storybook documentation and NPM package code with the version of `@wordpress/components` installed via `package.json` to confirm that the installed package contains the expected component(s)
   - WordPress components provide standard admin UI patterns
```

   - Once approved, follow existing Newspack component patterns (see [Component Patterns](#component-patterns))
   - Ensure it's reusable and follows WordPress design system principles
   - Consider if it should be a Newspack component or a wizard-specific component

## Design & layout at a glance

When building a screen, use the **spacing scale** (8px unit: 16, 24, 32, 48, 64) and prefer **VStack** for vertical stacks and **HStack** for related items in a row; use **Grid** for real multi-column layouts. Structure content as **SectionHeader → Card → ActionCard → controls**, with **Divider** between sections and primary actions in **`.newspack-buttons-card`**. Use the same **breakpoints** (e.g. 744px, 1128px) as existing components. Full detail: [Spacing scale](#spacing-scale-design-system), [Layout (HStack / VStack / Grid)](#layout-when-to-use-hstack-vstack-grid), [Responsive breakpoints](#responsive-breakpoints), [Visual hierarchy patterns](#visual-hierarchy-patterns), [Component states](#component-states). For code examples by context and wizard patterns, see [Usage](#usage).

## Available Components

### Layout Components

> - **`Card`** – Container for a logical block of content; use for grouping related settings. Default vertical margin is 32px so cards stack with consistent rhythm. Use `noBorder` when cards sit inside another card (e.g. ActionCard children).

Reviewer: Should we mention anything about the new `__experimentalCoreCard` and accompanying props? They almost act as a different component compared to the original.

Author: I was going to update this once your PR that introduces the Core card landed. Or maybe we can just add it to your PR?

Reviewer: That's fine, let's cover it in a separate PR after both this one and that one land!

- **`Divider`** – Use between logical sections (e.g. between ActionCards) to separate content without another card. Margins are 32px (64px at larger breakpoints) so spacing stays consistent with the rest of the layout.
- **`SectionHeader`** – Use to start a new section; pair with a short description so the section’s goal is clear. Top margin (64px) and bottom (32px) create clear separation from previous content and the section body.
- **`BoxContrast`** – High-contrast content box for emphasis.

### Form Components

- **`Button`** - Enhanced button component (wraps WordPress Button with routing support)
- **`TextControl`** - Text input with Newspack styling and required field support
- **`SelectControl`** - Select dropdown with grouped and button variants
- **`RadioControl`** - Radio button group control
- **`ColorPicker`** - Color selection component
- **`ImageUpload`** - Image upload and selection component
- **`FormTokenField`** - Token input field for tags/categories

> - **`AutocompleteTokenField`** - Autocomplete token field (can be used in blocks)

Reviewer: ```suggestion
- **`AutocompleteTokenField`** - Autocomplete token field (can be used in block editor)
```

- **`CustomSelectControl`** - Custom select dropdown component

### Content Components

- **`ActionCard`** – Use when one concept (e.g. a feature or setting) can be toggled on/off and may have extra content below. Internal padding (24px default; 16px/8px for isMedium/isSmall) and 24px between regions keep hierarchy clear; expandable content uses 24px top padding and 24px between siblings.
- **`ButtonCard`** – Card with button actions.
- **`Notice`** – Use for outcome feedback (success/error/warning) or short contextual messages. Vertical margin is 32px so notices don’t collide with cards; keep one primary message per area when possible.
- **`Waiting`** – Loading state indicator.
- **`ProgressBar`** – Progress indicator.
- **`Accordion`** – Collapsible content sections.
- **`StepsList`** / **`StepsListItem`** – Step-by-step list components.
- **`StyleCard`** – Style preview card.

### Wizard Components

- **`Wizard`** - Main wizard container with tabbed navigation and data fetching
- **`withWizard`** - Higher-order component for wizard screens (legacy pattern, class-based)
  - Provides plugin management, error handling, loading states
  - Used in older wizards like Setup Wizard

>   - Passes props: `wizardApiFetch`, `setError`, `isLoading`, `pluginRequirements`, etc.

Reviewer: ```suggestion
  - Passes props: `wizardApiFetch`, `setError`, `isLoading`, `pluginRequirements`, etc.
  - Do not use for new view components; use `Wizard` and/or `withWizardScreen` instead
```

  - Passes props: `renderPrimaryButton`, plus all original component props
- **`TabbedNavigation`** - Tab navigation component
- **`Footer`** - Wizard footer component
- **`Handoff`** - Handoff message component for external integrations
- **`HandoffMessage`** - Handoff message display component

### Plugin Management Components

- **`PluginInstaller`** - Plugin installation and activation component
- **`PluginToggle`** - Plugin enable/disable toggle
- **`PluginSettings`** - Plugin settings configuration component

### Utility Components

- **`Modal`** - Modal dialog component
- **`Popover`** - Popover component
- **`WebPreview`** - Web preview iframe component
- **`NewspackIcon`** - Newspack icon wrapper component
- **`InfoButton`** - Info button with tooltip
- **`GlobalNotices`** - Global notice system component

### Settings Components

- **`Settings`** - Settings container component
- **`SortableNewsletterListControl`** - Sortable newsletter list selector

### Utilities & Hooks

- **`hooks`** - Custom React hooks (e.g., `useObjectState`, `usePrompt`, `useOnClickOutside`)
- **`utils`** - Utility functions (e.g., `confirmAction`, color utilities)

> - **`Router`** - Proxied router import (use instead of direct `react-router-dom` import)

Reviewer: ```suggestion
- **`Router`** - Proxied React Router import (use instead of direct `react-router-dom` import)
  - Note: This package currently uses [React Router v5](https://v5.reactrouter.com/). Please refer to v5 documentation for API details.
```

**Example – Collections block:**
```jsx
import { PanelBody, TextControl } from '@wordpress/components';
import { AutocompleteTokenField } from '../../../../packages/components/src';

function InspectorPanel( { attributes, setAttributes } ) {
	return (
		<PanelBody title={ __( 'Settings', 'newspack-plugin' ) }>
			<TextControl
				label={ __( 'Title', 'newspack-plugin' ) }
				value={ attributes.title }
				onChange={ value => setAttributes( { title: value } ) }
			/>
			<AutocompleteTokenField
				label={ __( 'Categories', 'newspack-plugin' ) }
				value={ attributes.categories }
				onChange={ value => setAttributes( { categories: value } ) }
			/>
		</PanelBody>
	);
}
```

**Reference:** `src/blocks/collections/components/InspectorPanel.jsx`

**Why:** Blocks should match the block editor; WordPress components keep that consistency.

### Frontend (Reader-facing UI)

**Rule:** Don’t use Newspack components. Use Newspack UI (`src/newspack-ui/`) or theme components instead.

**Why:** This package is for backend/admin only; reader-facing UI uses other systems.

### Common patterns

> #### Wizard (modern)

Reviewer: We're adding some features to this component in #4474 that we might want to document here. The main additions:

- Items in the `sections` array can now receive their own props in a `props` array, and receive props from both the parent `Wizard` component and router props from the `Route` component rendering each section
- `Wizard` components can receive `headerData` data to set the following via Redux store. These can be set using the `setHeaderData` dispatcher exported by the Wizard store:
  - `sectionName`: a string that will be rendered as a breadcrumb item in the Wizard header bar
  - `sectionTitle`: a string that will be used to render a styled `SectionHeader` component at the top of the section content.
  - `backNav`: a URL string. If provided, the `SectionHeader` component will also render a small back button link to navigate to the provided URL.
  - `badges`: an array of badge data to render one or more `Badge` components alongside the `sectionTitle`. Config is as follows:

```ts
type BadgeProps = {
	text: string; // A string to render as text inside the Badge
	level?: 'default' | 'info' | 'success' | 'warning' | 'error'; // A string to indicate which style the Badge should display
};
```

  - `actions`: an array of actions that will be rendered on the right-hand side of the Wizard header bar. Config is as follows:

```ts
type HeaderAction = {
	type: 'primary' | 'secondary' | 'more'; // A string indicating how the action should be represented in UI. `primary` and `secondary` = a `Button` Newspack component with the corresponding `variant` prop. `more`: as a `MenuItem` inside a `DropdownMenu` component.
	label: string; // The human-readable label for the UI component.
	icon: React.ReactNode; // An optional icon to display in the UI component.
	action: () => void; // A function for the UI component to call.
	disabled?: boolean; // If true, the UI component will be rendered in a disabled state.
	destructive?: boolean; // If true, the UI component will be rendered with a destructive style.
};
```

**Example of setting header data using the dispatcher:**

```jsx
const { setHeaderData } = useDispatch( WIZARD_STORE_NAMESPACE );
setHeaderData( {
	backNav: '#/content-gates', // Can use relative or absolute paths
	sectionName: __( 'Breadcrumb title', 'newspack-plugin' ),
	sectionTitle: __( 'Section header title', 'newspack-plugin' ),
} );
```

This may change depending on how that PR goes, but maybe worth adding and keeping up to date?

Reviewer: Similar to the thread about `__experimentalCoreCard` [here](https://github.com/Automattic/newspack-plugin/pull/4493#discussion_r2819569710), let's cover this in a separate PR after this PR and the other one land.

/**
 * WordPress dependencies
 */
import { __ } from '@wordpress/i18n';
import { Fragment } from '@wordpress/element';

/**
 * Internal dependencies
 */
import { GlobalNotices, Notice, Wizard } from '../../../../../packages/components/src';
import sections from './sections';

function Dashboard() {
	return (
		<Fragment>
			<GlobalNotices />
			<Wizard
				headerText={ __( 'Newspack / Dashboard', 'newspack-plugin' ) }
				sections={ sections }
				renderAboveSections={ () => (
					<>
						<BrandHeader />
						<SiteStatuses />
					</>
				) }
			/>
		</Fragment>
	);
}
```

### Using withWizardScreen HOC (Modern Pattern)

Use `withWizardScreen` for modern wizard screens that need header, navigation, and button actions:

```jsx
/**
 * Internal dependencies
 */
import { withWizardScreen, ActionCard, Button, Card } from '../../../../../packages/components/src';

export default withWizardScreen( ( { config, updateConfig, saveConfig, renderPrimaryButton } ) => {
	return (
		<Card noBorder>
			<ActionCard
				title={ __( 'Feature', 'newspack-plugin' ) }
				toggleChecked={ config.enabled }
				toggleOnChange={ value => updateConfig( 'enabled', value ) }
			/>
		</Card>
	);
} );
```

> **When to use:** Modern wizard screens (Audience Management, Settings sections, etc.)

Reviewer: ```suggestion
**When to use:** Modern wizard screens (Audience Management, Settings sections, etc.). Use when you need more precise control over the layout and routing structure in a wizard view than the `Wizard` component can provide.
```

### Using withWizard HOC (Legacy Pattern)

Use `withWizard` for legacy wizard screens that need plugin management and error handling:

```jsx
/**
 * Internal dependencies
 */
import { withWizard, Card, PluginInstaller } from '../../../../../packages/components/src';

function MyWizardScreen( { wizardApiFetch, setError, isLoading, pluginRequirements } ) {
	return (
		<>
			{ pluginRequirements }
			<Card>
				{/* Wizard content */}
			</Card>
		</>
	);
}

export default withWizard( MyWizardScreen, [ 'required-plugin-slug' ] );
```

> **When to use:** Legacy wizards (Setup Wizard, older wizard implementations)

Reviewer: ```suggestion
**When to use:** Legacy wizards (Setup Wizard, older wizard implementations). Do not use for new wizards.
```

### Using Hooks and Utilities

```jsx
/**
 * Internal dependencies
 */
import { hooks, utils } from '../../../../../packages/components/src';

function MyComponent() {
	// useObjectState hook for managing object state
	const [ data, setData ] = hooks.useObjectState( {
		field1: '',
		field2: false,
	} );
	
	// Update nested field
	setData( { field1: 'new value' } );
	
	// Confirm action utility
	const handleDelete = () => {
		if ( utils.confirmAction( __( 'Are you sure?', 'newspack-plugin' ) ) ) {
			// Proceed with deletion
		}
	};
}
```

> ### Using AutocompleteTokenField in Blocks

Reviewer: ```suggestion
### Using AutocompleteTokenField in the Block Editor
```

```jsx
/**
 * WordPress dependencies
 */
import { PanelBody } from '@wordpress/components';

/**
 * Internal dependencies
 */
import { AutocompleteTokenField } from '../../../../packages/components/src';

function InspectorPanel( { attributes, setAttributes } ) {
	return (
		<PanelBody>
			<AutocompleteTokenField
				label={ __( 'Categories', 'newspack-plugin' ) }
				value={ attributes.categories }
				onChange={ value => setAttributes( { categories: value } ) }
				suggestions={ [ 'News', 'Opinion', 'Sports' ] }
			/>
		</PanelBody>
	);
}
```

## Common WordPress Components Used Alongside Newspack Components

When Newspack components don't provide what you need, use these WordPress components. The list below reflects actual usage across the codebase (wizards, blocks, packages/components, other scripts).

### Form controls

- **`CheckboxControl`** – Checkbox input (very common in wizards and blocks)
- **`ToggleControl`** – Toggle switch (common in settings and content gates)
- **`RangeControl`** – Numeric range slider

> - **`TextControl`** – Single-line text input

Reviewer: There's a Newspack Component wrapper for this—should we prefer that one instead?

Author: What's the advantage of the Newspack version?

Reviewer: [See above thread for details](https://github.com/Automattic/newspack-plugin/pull/4493#discussion_r2842176621)

- **`TextareaControl`** – Multi-line text input

> - **`SelectControl`** – Select dropdown

Reviewer: There's a Newspack Component wrapper for this—should we prefer that one instead?

Author: - I don't see a RangeControl
- TextControl seems unnecessary (the classes we add can easily be added as generic classes)
- We don't have TextareaControl
- We should probably use our SelectControl? I'm not sure about the use case.

Reviewer: > I don't see a RangeControl

To clarify, this is the list of core `@wordpress/components` that are commonly used when a corresponding Newspack component doesn't exist. So `CheckboxControl`, `ToggleControl`, `RangeControl`, `TextareaControl`, `__experimentalNumberControl as NumberControl`, and `__experimentalToggleGroupControl`** / **`__experimentalToggleGroupControlOption` all make sense to include here.

What I'm asking is whether we should omit `TextControl`, `SelectControl`, and `FormTokenField` from the list because there ARE Newspack components covering these.

Author: Ah I see. What are the differences between Core and ours?

Reviewer: `FormTokenField` is mostly just a wrapper for the core component, but also supports a `description` prop that shows a description like other core controls, which the core component does not for some reason.

`SelectControl` is a wrapper for core's `SelectControl`, but also `GroupedSelectControl` and `ButtonGroupControl` based on the presence of `optgroups` or `buttonOptions` props.

`ToggleControl` has some different styles and additional CSS classes compared to the core component but is mostly a wrapper for the core component.

Author: I can't find any examples for `ToggleControl` using Newspack's. I wonder if we should just remove this component...

I see mixed uses of `FormTokenField`. I think we need to standardise this... Probably Newspack's since we have the help text/description

And I also see mixed uses of `SelectControl` 🤔

Reviewer: Yes, in that case I think we should document what our desired standard is and not worry too much about the existing inconsistencies.

Author: Ok so:

* In https://github.com/Automattic/newspack-plugin/pull/4493/commits/67698f6c01f2df8cbc75302c8f58187a253e4378 I added preference for our FormTokenField
* In https://github.com/Automattic/newspack-plugin/pull/4493/commits/1b36c22d0f794128d5f4ea67aed6baae9ecc5784 I removed our SelectControl. Group is now supported by the Core component and the Button-style should be avoided and we should use ToggleGroupControl anyway

Author: Note: 

I think we should remove/delete our `SelectControl` and `ToggleControl`

Reviewer: Thanks for clarifying! To make this intention official, can we add a deprecation note/comment to the `SelectControl` and `ToggleControl` component package files? We don't have to deprecate them now/anytime soon, but at least we can codify the intention so we know to stop using them for new features.

Author: Done: https://github.com/Automattic/newspack-plugin/pull/4493/commits/c65deb47bea58a4ec6f7d22eaf325a3a89292cb9

We actually don't have a `ToggleControl` 😂

> - **`FormTokenField`** – Token/tag input (e.g. categories, tags)

Reviewer: There's a Newspack Component wrapper for this—should we prefer that one instead?

Author: See https://github.com/Automattic/newspack-plugin/pull/4493#discussion_r2841436356

	__experimentalNumberControl as NumberControl, // eslint-disable-line @wordpress/no-unsafe-wp-apis
	__experimentalVStack as VStack, // eslint-disable-line @wordpress/no-unsafe-wp-apis
	ExternalLink,
} from '@wordpress/components';
```

**Example – form controls and layout:**
```jsx
import {
	CheckboxControl,
	ExternalLink,
	__experimentalHStack as HStack, // eslint-disable-line @wordpress/no-unsafe-wp-apis
	RangeControl,
} from '@wordpress/components';
import { ActionCard, Card } from '../../../../../packages/components/src';

<ActionCard title="Settings">
	<HStack>
		<RangeControl
			label={ __( 'Value', 'newspack-plugin' ) }
			value={ config.value }
			onChange={ value => updateConfig( 'value', value ) }
		/>
		<CheckboxControl
			label={ __( 'Enable', 'newspack-plugin' ) }
			checked={ config.enabled }
			onChange={ value => updateConfig( 'enabled', value ) }
		/>
	</HStack>
</ActionCard>
```

## Styling

Newspack components use SCSS with BEM-ish naming conventions and a consistent spacing scale so layouts stay visually consistent.

### Naming and structure

- **Prefix:** `newspack-` (e.g. `.newspack-card`, `.newspack-button`)

> - **Modifiers:** Use `--` for modifiers (e.g. `.newspack-card--no-border`)

Reviewer: We also use `__` for hierarchy as [BEM describes](https://getbem.com/naming/#element), but we're kind of inconsistent about it. In some components `__` is used as if it were a `--` modifier, and vice versa. Still, maybe we should document both and their intended usage?

```suggestion
- **Modifiers:** Use `--` for modifiers (e.g. `.newspack-card--no-border`)
- **Elements**: Use `__` for elements that are part of a larger block-level component (e.g. `.newspack-card__header-content`)
```

**Example - ActionCard Component:**
```jsx
// Custom component with toggle, actions, expandable content, etc.
const ActionCard = ( {
	title,
	description,
	toggleChecked,
	toggleOnChange,
	children,
	// ... many other props
} ) => {
	// Custom implementation
};
```

## Component Development Guidelines

When creating new components:

1. **Check if WordPress component exists** – Consider wrapping/extending WordPress components first.
2. **Align with design** – For new patterns or layouts (e.g. a new card style or wizard step), get designer review before implementing so spacing, hierarchy, and component choice match the design system.
3. **Follow import patterns** – Use the standard import order with JSDoc comments.
4. **Use TypeScript** – Prefer `.tsx` for new components (codebase is migrating to TypeScript).
5. **Add PropTypes or TypeScript types** – Document component props.
6. **Include styles** – Add component-specific styles in `style.scss`; use the [spacing scale](#spacing-scale-design-system) (8px multiples: 16, 24, 32, 48, 64) so new components match Card, ActionCard, and Grid.
7. **Follow naming conventions** – Use BEM-ish naming with `newspack-` prefix.

> 8. **Use WordPress design system** – Leverage WordPress colors and the same spacing values as existing components.

Reviewer: Should we mention something about always translating static strings?

Author: Done: https://github.com/Automattic/newspack-plugin/pull/4493/commits/be35297ab8cc4ea6f664596486035ce3e08eb7f8

9. **Export from index.js** – Add component to `packages/components/src/index.js`.

> 10. **Document usage** – Add JSDoc comments and update this guide.

Reviewer: ```suggestion
10. **Document usage** – Add JSDoc comments and update this guide.
11. **Components Demo** - If the component is complex or would benefit from demo examples, add it to the [Components Demo page](#testing).
```

> 

Reviewer: Might be worth mentioning since we have examples of both in our codebase:

### Prefer functional components over class-based components

Modern React APIs [prefer functional components over class components](https://react.dev/reference/react/Component), even though both are currently still supported. 

Some older Newspack components are still class-based, while newer components are functional. When creating new Newspack components, use functional components instead of class components.

```jsx
// ✅ CORRECT - example from the Divider component
/**
 * Divider
 */

/**
 * Internal dependencies
 */
import './style.scss';

/**
 * External dependencies
 */
import classNames from 'classnames';

const Divider = ( { alignment = 'none', className = undefined, marginBottom = 64, marginTop = 64, variant = 'default', ...otherProps } ) => {
	const classes = classNames(
		'newspack-divider',
		className,
		alignment && `newspack-divider--alignment-${ alignment }`,
		variant && `newspack-divider--variant-${ variant }`
	);

	const style = {
		'--divider-margin-bottom': typeof marginBottom === 'number' ? `${ marginBottom }px` : marginBottom,
		'--divider-margin-top': typeof marginTop === 'number' ? `${ marginTop }px` : marginTop,
	};

	return <hr className={ classes } style={ style } { ...otherProps } />;
};

export default Divider;
```

```jsx
// ❌ WRONG – Example from the older `Popover` component
/**
 * WordPress dependencies
 */
import { Popover as BaseComponent } from '@wordpress/components';
import { Component } from '@wordpress/element';

/**
 * Internal dependencies
 */
import './style.scss';

/**
 * External dependencies
 */
import classnames from 'classnames';

/**
 * Popover
 */
class Popover extends Component {
	/**
	 * Render
	 */
	render() {
		const { className, padding, ...otherProps } = this.props;
		const classes = classnames( 'newspack-popover', padding && 'newspack-popover__padding-' + padding, className );
		return <BaseComponent className={ classes } { ...otherProps } />;
	}
}

Popover.defaultProps = {
	padding: false,
};

export default Popover;
```

Author: Added this: https://github.com/Automattic/newspack-plugin/pull/4493/commits/97a775a5ed0962b26e55a598130d9cf2445ca6fe
